[SPARK-24923][SQL] Implement v2 CreateTableAsSelect

rdblue opened the pull request:
## What changes were proposed in this pull request?

This adds a v2 implementation for CTAS queries

* Update the SQL parser to parse CREATE queries using multi-part identifiers
* Update `CheckAnalysis` to validate partitioning references with the CTAS query schema
* Add `CreateTableAsSelect` v2 logical plan and `CreateTableAsSelectExec` v2 physical plan
* Update create conversion from `CreateTableAsSelectStatement` to support the new v2 logical plan
* Update `DataSourceV2Strategy` to convert v2 CTAS logical plan to the new physical plan
* Add `findNestedField` to `StructType` to support reference validation

## How was this patch tested?

We have been running these changes in production for several months. Also:

* Add a test suite `CreateTablePartitioningValidationSuite` for new analysis checks
* Add a test suite for v2 SQL, `DataSourceV2SQLSuite`
* Update catalyst `DDLParserSuite` to use multi-part identifiers (`Seq[String]`)
* Add test cases to `PlanResolutionSuite` for v2 CTAS: known catalog and v2 source implementation

rdblue committed:
Implement v2 CreateTableAsSelect.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10567/
Test PASSed.

SparkQA commented:
**[Test build #105292 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105292/testReport)** for PR 24570 at commit [`b13a8e2`](https://github.com/apache/spark/commit/b13a8e27c38638b3b7487565fd2d1b2ed7d54a9e).

SparkQA commented:
**[Test build #105292 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105292/testReport)** for PR 24570 at commit [`b13a8e2`](https://github.com/apache/spark/commit/b13a8e27c38638b3b7487565fd2d1b2ed7d54a9e).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTableAsSelect(`
  * `case class DataSourceResolution(`
  * `case class CreateTableAsSelectExec(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105292/
Test FAILed.

rdblue committed:
Add v2 SQL test suite.

rdblue committed:
Update HiveSessionStateBuilder.

SparkQA commented:
**[Test build #105296 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105296/testReport)** for PR 24570 at commit [`a22c335`](https://github.com/apache/spark/commit/a22c33578a0637fb093f6eeb4c00e8e09b7220b2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10570/
Test PASSed.

SparkQA commented:
**[Test build #105297 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105297/testReport)** for PR 24570 at commit [`cdf3805`](https://github.com/apache/spark/commit/cdf38051fd4ab85277c7e285acf92ed8ba3b8c1b).

SparkQA commented:
**[Test build #105296 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105296/testReport)** for PR 24570 at commit [`a22c335`](https://github.com/apache/spark/commit/a22c33578a0637fb093f6eeb4c00e8e09b7220b2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105296/
Test PASSed.

SparkQA commented:
**[Test build #105297 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105297/testReport)** for PR 24570 at commit [`cdf3805`](https://github.com/apache/spark/commit/cdf38051fd4ab85277c7e285acf92ed8ba3b8c1b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105297/
Test PASSed.

rdblue commented:
@mccheah, @cloud-fan, here's a v2 implementation for CTAS. Please review if you have time. Thanks!

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala:
```diff
@@ -46,14 +57,22 @@ case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with Ca
+    case create: CreateTableAsSelectStatement =>
+      // the provider was not a v1 source, convert to a v2 plan
+      val CatalogObjectIdentifier(maybeCatalog, identifier) = create.tableName
+      val catalog = maybeCatalog.orElse(defaultCatalog)
```
The ability to change the default catalog can be done later. I'd like to add it when we migrate hive catalog to `TableCatalog` API, so that the default can be hive catalog.

rdblue replied:
Sure. I'll remove this `orElse` and disable the related tests.

rdblue replied:
This is in #24594.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala:
```diff
@@ -46,14 +57,22 @@ case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with Ca
 
       CreateTable(tableDesc, mode, Some(query))
+
+    case create: CreateTableAsSelectStatement =>
```
shall we implement CREATE TABLE as well?

rdblue replied:
I thought this PR was getting a little large, so I was going to do create table in a separate one. I can add it to this one if you'd prefer.

rdblue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -296,6 +298,21 @@ trait CheckAnalysis extends PredicateHelper {
+          case CreateTableAsSelect(_, _, partitioning, query, _, _, _) =>
+            val references = partitioning.flatMap(_.references).toSet
+            val badReferences = references.map(_.fieldNames).flatMap { column =>
+              query.schema.findNestedField(column).map(_.dataType) match {
```
I'll remove it, thanks.

rdblue committed:
Minor updates after comments.

rdblue committed:
Add missing test suite.

rdblue committed:
Remove default catalog.

rdblue commented:
@cloud-fan, I've updated this PR and fixed your review comments, so please have another look. Also, I added a test suite that I accidentally didn't add, `DataSourceV2SQLSuite`. That is where I will be putting end-to-end test cases that use in-memory tables to validate the physical plans.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10629/
Test PASSed.

SparkQA commented:
**[Test build #105364 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105364/testReport)** for PR 24570 at commit [`b19a70d`](https://github.com/apache/spark/commit/b19a70dee4bf6b36ed8ed6c835a5729c2e171a83).

SparkQA commented:
**[Test build #105364 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105364/testReport)** for PR 24570 at commit [`b19a70d`](https://github.com/apache/spark/commit/b19a70dee4bf6b36ed8ed6c835a5729c2e171a83).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105364/
Test PASSed.

mccheah commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala:
```diff
@@ -112,4 +129,41 @@ case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with Ca
+
+    // convert USING, LOCATION, and COMMENT clauses to table properties
+    properties += ("provider" -> ctas.provider)
+    ctas.comment.map(text => properties += ("comment" -> text))
```
What happens if these overwrite table properties of the same type in the properties part of the SQL statement itself? Do these overwrite properties of the same name? Are these special property keys documented anywhere?

cloud-fan replied:
This is a good point. When saving Spark tables to Hive metastore, we need to store some spark specific information in the table properties. And we always use `spark.sql.` as the prefix for the property keys.

Shall we follow it and add a prefix here as well?

rdblue replied:
I've added validations so that the properties and the clauses cannot both be used. Setting "comment" and using a COMMENT clause will result in an `AnalysisException`.

Passing these as well-known properties was included in the SPIP, but exactly which properties are used should be documented. I'll open a documentation issues and add it as a blocker for the 3.0 release.

If we want to define a prefix for the clauses that are passed as properties, that sounds fine to me. What is a reasonable prefix?

rdblue replied:
With the latest changes, conflicting properties are no longer allowed.

Also, I opened https://issues.apache.org/jira/browse/SPARK-27708 to track documentation.

mccheah commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -402,6 +404,35 @@ trait V2WriteCommand extends Command {
+    // that the columns referenced by the table's partitioning exist in the query schema
+    val references = partitioning.flatMap(_.references).toSet
+    references.map(_.fieldNames).forall { column =>
+      query.schema.findNestedField(column) match {
```
Can we just check on `isDefined` on the returned option?

cloud-fan replied:
ah good catch, it should be `query.schema.findNestedField(column).isDefined`

mccheah commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -2031,7 +2031,8 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
       operationNotAllowed("CREATE TEMPORARY TABLE ... IF NOT EXISTS", ctx)
     }
-    (visitTableIdentifier(ctx.tableIdentifier), temporary, ifNotExists, ctx.EXTERNAL != null)
+    val multipartIdentifier: Seq[String] = ctx.multipartIdentifier.parts.asScala.map(_.getText)
```
Is there a real need to declare the type?

rdblue replied:
Yes, the compiler was acting strangely and I needed to add it.

rdblue replied:
I just took another look and it is no longer needed. Must have been a problem while I hadn't updated everything. I've fixed it.

mccheah commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -47,6 +51,60 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
+              .withQueryId(UUID.randomUUID().toString)
+          val batchWrite = builder match {
+            case supportsSaveMode: SupportsSaveMode =>
+              supportsSaveMode.mode(SaveMode.Append).buildForBatch()
```
I'm not sure why we have to specifically tell the writer to use append mode. Can you elaborate? I think I'm missing something. It would be simpler to remove this branch entirely if possible.

cloud-fan replied:
I think `SupportsSaveMode` is a hack in `TableProvider` only. Seems we don't need to deal with it for `TableCatalog`. Anyway it will be removed soon.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/CreateTablePartitioningValidationSuite.scala:
```diff
@@ -0,0 +1,153 @@
+    assert(!plan.resolved)
+    assertAnalysisError(plan, Seq(
+      "Invalid partitioning",
+      "does_not_exist.z is missing or is in a map or array"))
```
Not a blocker, but we can improve the error message in the future. It's better to let users know which column/field is missing. For example, `a.b`, it's possible that column `a` exists but `a` is not a struct or it doesn't have a `b` field.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala:
```diff
@@ -46,14 +55,22 @@ case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with Ca
+      // the provider was not a v1 source, convert to a v2 plan
+      val CatalogObjectIdentifier(maybeCatalog, identifier) = create.tableName
+      val catalog = maybeCatalog
+          .getOrElse(throw new AnalysisException("Default catalog is not set"))
```
nit: `catalog is not specified in the table identifier`.

rdblue replied:
No, this is the correct error message. When no explicit catalog is used, Spark should use the default catalog. When there is no catalog, the problem is that there was no default set.

This is why the default catalog was originally part of this PR. The errors depend on it.

rdblue replied:
I'm updating the error message to include both: `No catalog specified for table ... and no default catalog is set`

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala:
```diff
@@ -112,4 +129,41 @@ case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with Ca
+      ctas: CreateTableAsSelectStatement): CreateTableAsSelect = {
+    if (ctas.options.contains("path") && ctas.location.isDefined) {
+      throw new AnalysisException(
+        "LOCATION and 'path' in OPTIONS are both used to indicate the custom table path, " +
```
just for curiosity, does v1 fail this case too?

cloud-fan replied:
and shall we do the same check for `provider` and `comment`?

rdblue replied:
Yes, it does fail. And I can add checks for provider and comment.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala:
```diff
@@ -0,0 +1,164 @@
+
+  before {
+    spark.conf.set("spark.sql.catalog.testcat", classOf[TestInMemoryTableCatalog].getName)
+    spark.conf.set("spark.sql.default.catalog", "testcat")
```
I think it has been moved to another PR?

rdblue replied:
It has. Does this need to be removed? I made the tests that rely on it ignored, so we should just need to revert that. I'd rather keep it so we don't have to figure out why those tests are failing when the other PR is merged.

rdblue committed:
More changes from the review.

rdblue committed:
Remove unnecessary declared type.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10649/
Test PASSed.

SparkQA commented:
**[Test build #105387 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105387/testReport)** for PR 24570 at commit [`99ebc00`](https://github.com/apache/spark/commit/99ebc001d2563ad579e5e9a0211606cec4e396be).

SparkQA commented:
**[Test build #105387 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105387/testReport)** for PR 24570 at commit [`99ebc00`](https://github.com/apache/spark/commit/99ebc001d2563ad579e5e9a0211606cec4e396be).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105387/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

rdblue commented:
Thanks for reviewing, @cloud-fan and @mccheah! I'll get the next few PRs posted.
